stubs/docutils/docutils/parsers/rst/roles.pyi

```diff
@@ -1,20 +1,131 @@
-from _typeshed import Incomplete
 from collections.abc import Callable
 from typing import Any
 from typing_extensions import TypeAlias
 
-import docutils.nodes
 import docutils.parsers.rst.states
+from docutils import nodes
 from docutils.languages import _LanguageModule
+from docutils.nodes import Node, TextElement
+from docutils.parsers.rst.states import Inliner
 from docutils.utils import Reporter, SystemMessage
 
+DEFAULT_INTERPRETED_ROLE: str
+
 _RoleFn: TypeAlias = Callable[
     [str, str, str, int, docutils.parsers.rst.states.Inliner, dict[str, Any], list[str]],
-    tuple[list[docutils.nodes.reference], list[docutils.nodes.reference]],
+    tuple[list[nodes.reference], list[nodes.reference]],
 ]
 
+def register_canonical_role(name: str, role_fn: _RoleFn) -> None: ...
 def register_local_role(name: str, role_fn: _RoleFn) -> None: ...
 def role(
     role_name: str, language_module: _LanguageModule, lineno: int, reporter: Reporter
 ) -> tuple[_RoleFn | None, list[SystemMessage]]: ...
-def __getattr__(name: str) -> Incomplete: ...
+def set_implicit_options(role_fn: _RoleFn) -> None: ...
+def register_generic_role(canonical_name: str, node_class: Node) -> None: ...
+
+class GenericRole:
+    name: str
+    node_class: Node
+    def __init__(self, role_name: str, node_class: type[TextElement]) -> None: ...
+    def __call__(
+        self,
+        role: str,
+        rawtext: str,
+        text: str,
+        lineno: int,
+        inliner: Inliner,
+        options: dict[str, Any] | None = None,
```

Comment:
Prefer using `Mapping` if possible. Similarly, instead of `list`, it's probably better to use `Sequence` for many of these.

Reply:
i've replaced most of them. I kept a couple in return value position, and in 'normalized_role_options' since it returns a dict, but may forward the input options directly

Reply:
Oh I see, sorry. Argument annotations in callbacks are one place where it's good to be as narrow as possible, because that means more callbacks are accepted (since they behave contravariantly). I'm guessing the options dict gets passed to the RoleFn callbacks? It might be best to stick with the dict/list types to make sphinx's life simpler.